### PR TITLE
[WIP] Conversion plugin for Nikola's #2761: conversion plugin for pre-v8 tag-style metadata

### PR DESCRIPTION
**Pull request by felixfontein**
Companions #3068.

### CODE REVIEW

**Comment by felixfontein** [comment]
Ok, first working version.

**Comment by felixfontein** [comment]
(requires getnikola/nikola#3068 to work)

File: v8/upgrade_metadata_v8/upgrade_metadata_v8.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+                            continue
+
+                        # Recombine metadata with post text if necessary, and write back to file
+                        meta_str = extractor.write_metadata(meta)
+                        print(meta_str)
+                        if not is_two_file:
+                            if extractor.name == 'nikola':
+                                final_str = meta_str + '\n\n' + content_str
+                            elif extractor.name == 'yaml':
+                                final_str = meta_str + '\n---\n' + content_str
```
Comment (Kwpolska) [inline v8/upgrade_metadata_v8/upgrade_metadata_v8.py]: Is there a way not to hardcode this?

File: v8/upgrade_metadata_v8/upgrade_metadata_v8.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                        }
+                        for foreign, ours in self.site.config.get('METADATA_MAPPING', {}).get(extractor.map_from, {}).items():
+                            if ours in sources:
+                                sources[ours] = foreign
+                        for meta_key, hook in self.site.config.get('METADATA_VALUE_MAPPING', {}).get(extractor.map_from, {}).items():
+                            if meta_key in sources.values():
+                                L.warn('Cannot convert {0} (language {1}): a metadata value mapping is defined for "{2}"!'.format(fname, lang, meta_key))
+
+                        # Update metadata
+                        updated = False
```
Comment (Kwpolska) [inline v8/upgrade_metadata_v8/upgrade_metadata_v8.py]: Feature request:

if `section` is present, but `category` isn’t: set `category` to `section` value and `remove` section  
if both are present: show a warning (as is done on the `nuke-sections` branch of the main repo)

**Comment by ralsina** [comment]
+1, not familiar with the section killing, so going with Kw's opinion there.

File: v8/upgrade_metadata_v8/upgrade_metadata_v8.py (hunk 1 above)
Comment (felixfontein) [reply v8/upgrade_metadata_v8/upgrade_metadata_v8.py]: We could extend the metadata extractors so that they offer a recombine function which does this. That would avoid this. If this is ok for you, I'll start a PR for that.

Comment (felixfontein) [reply v8/upgrade_metadata_v8/upgrade_metadata_v8.py]: Aaah, I missed that part :)

File: v8/upgrade_metadata_v8/upgrade_metadata_v8.py (hunk 2 above)
Comment (felixfontein) [reply v8/upgrade_metadata_v8/upgrade_metadata_v8.py]: Implemented in 685be9cb24cf13a4b9978baefa2e01c44ef2200d.

**Comment by felixfontein** [comment]
Now the tests also no longer fail, thanks to 42d6e2b38bf01976b44af29480f4e1c09f42c230.

File: v8/upgrade_metadata_v8/upgrade_metadata_v8.py (hunk 1 above)
Comment (Kwpolska) [reply v8/upgrade_metadata_v8/upgrade_metadata_v8.py]: You could fix 2\. by checking if the format is reST.

**Comment by felixfontein** [comment]
@Kwpolska how do you want to proceed with this one?

**Comment by felixfontein** [comment]
(I guess for a v8 beta, this one should be available.)

File: v8/upgrade_metadata_v8/upgrade_metadata_v8.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            if post.has_oldstyle_metadata_tags:
+                flag = True
+            for lang in self.site.config['TRANSLATIONS'].keys():
+                if 'section' in post.meta[lang]:
+                    flag = True
+            if flag:
+                flagged.append(post)
+        if flagged:
+            if len(flagged) == 1:
+                L.info('1 post (and/or its translations) contains old-style metadata or have section metadata:')
```
Comment (Kwpolska) [inline v8/upgrade_metadata_v8/upgrade_metadata_v8.py]: you mixed up have/has in both cases

File: v8/upgrade_metadata_v8/upgrade_metadata_v8.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                            meta[sources['tags']] = ', '.join(tags)
+
+                        if not updated:
+                            # Nothing to do (but successful)!
+                            converted = True
+                            continue
+
+                        # Recombine metadata with post text if necessary, and write back to file
+                        meta_str = utils.write_metadata(meta, metadata_format=extractor.name, compiler=post.compiler,
+                                                        comment_wrap=True, site=self.site)
```
Comment (Kwpolska) [inline v8/upgrade_metadata_v8/upgrade_metadata_v8.py]: `comment_wrap=(post.compiler.name != 'rest')`

File: v8/upgrade_metadata_v8/upgrade_metadata_v8.py (hunk 3 above)
Comment (felixfontein) [reply v8/upgrade_metadata_v8/upgrade_metadata_v8.py]: Thanks! Fixed.

**Comment by Kwpolska** [comment]
Did you test this with YAML or other tricky edge cases?

**Comment by felixfontein** [comment]
Tags as list work now. What doesn't work are two-file posts, it seems. They don't work, because no extractor is stored for them. `get_metadata_from_meta_file` simply doesn't return it. Is this intentional?

**Comment by felixfontein** [comment]
(YAML and TOML work fine, at least outside two-files.)

**Comment by Kwpolska** [comment]
Fixed on master (c2044a2e)

**Comment by Kwpolska** [comment]
Fixed on master (getnikola/nikola@c2044a2e)

**Comment by felixfontein** [comment]
LOL, you were faster. I just pushed a branch for that :)

**Comment by felixfontein** [comment]
Now two-file posts work as well. `METADATA_MAPPING` support also seems to work.

**Comment by felixfontein** [comment]
I think it should be fine now.

**Comment by Kwpolska** [comment]
There were two bugs I found while fixing `chriswarrick.com`:

1. saving as UTF-8 with BOM is evil
2. posts without translations crash:

```pytb
Traceback (most recent call last):
  File "/Users/kwpolska/virtualenvs/nikola/lib/python3.6/site-packages/doit/doit_cmd.py", line 172, in run
    return command.parse_execute(args)
  File "/Users/kwpolska/virtualenvs/nikola/lib/python3.6/site-packages/doit/cmd_base.py", line 127, in parse_execute
    return self.execute(params, args)
  File "/Users/kwpolska/git/nikola/nikola/plugin_categories.py", line 147, in execute
    return self._execute(options, args)
  File "/Users/kwpolska/website/plugins/upgrade_metadata_v8/upgrade_metadata_v8.py", line 105, in _execute
    with io.open(fname, "r", encoding="utf-8-sig") as meta_file:
FileNotFoundError: [Errno 2] No such file or directory: 'posts/2017/09/02/spawning-subprocesses-smartly-and-securely.pl.rst'
```

I fixed both in commit f79b5c5.

**Comment by felixfontein** [comment]
Ah, that's what `-sig` is for. I copied the `open` command from `post.py` (and changed it from read to write), and forgot to think about the encoding specified there... Sorry for that!

Thanks for fixing this.
